services/graph/pkg/service/v0/users_filter.go

```diff
@@ -126,22 +165,136 @@ func (g Graph) applyMemberOfEq(ctx context.Context, req *godata.GoDataRequest, n
+		return users, invalidFilterError()
+	}
+
+	// We only support the 'eq' expression for now
+	if nodes[1].Token.Type != godata.ExpressionTokenLogical || nodes[1].Token.Value != "eq" {
+		return users, unsupportedFilterError()
+	}
+	return g.applyAppRoleAssignmentEq(ctx, req, nodes[1].Children)
+}
 
+func (g Graph) applyAppRoleAssignmentEq(ctx context.Context, req *godata.GoDataRequest, nodes []*godata.ParseNode) (users []*libregraph.User, err error) {
+	logger := g.logger.SubloggerWithRequestID(ctx)
+	if len(nodes) != 2 {
+		return users, invalidFilterError()
+	}
+
+	if nodes[0].Token.Type != godata.ExpressionTokenNav {
+		return users, invalidFilterError()
+	}
+
+	if len(nodes[0].Children) != 2 {
+		return users, invalidFilterError()
+	}
+
+	if nodes[0].Children[1].Token.Value == appRoleID {
+		filterValue, err := g.getUUIDTokenValue(ctx, nodes[1])
+		if err != nil {
+			return users, unsupportedFilterError()
+		}
+
+		logger.Debug().Str("property", nodes[0].Children[1].Token.Value).Str("value", filterValue).Msg("Filtering appRoleAssignments by appRoleId")
+		if users, err = g.identityBackend.GetUsers(ctx, req); err != nil {
+			return users, err
+		}
+
+		return g.filterUsersByAppRoleID(ctx, filterValue, users)
+	}
+	return users, unsupportedFilterError()
+}
+
+func (g Graph) filterUsersByAppRoleID(ctx context.Context, id string, users []*libregraph.User) ([]*libregraph.User, error) {
+	// We're using a map for the results here, in order to avoid returning
+	// a user twice. The settings API, still has an issue that causes it to
+	// duplicate some assignments on restart:
+	// https://github.com/owncloud/ocis/issues/3432
+	resultUsersMap := make(map[string]*libregraph.User, len(users))
+	for _, user := range users {
+		assignments, err := g.fetchAppRoleAssignments(ctx, user.GetId())
+		if err != nil {
+			return users, err
+		}
+		for _, assignment := range assignments {
+			if assignment.GetAppRoleId() == id {
+				if _, ok := resultUsersMap[user.GetId()]; !ok {
+					resultUsersMap[user.GetId()] = user
+				}
+			}
+		}
+	}
+	resultUsers := make([]*libregraph.User, 0, len(resultUsersMap))
+	for _, user := range resultUsersMap {
+		resultUsers = append(resultUsers, user)
+	}
+	return resultUsers, nil
+}
+
+func (g Graph) getUUIDTokenValue(ctx context.Context, node *godata.ParseNode) (string, error) {
+	var value string
+	switch node.Token.Type {
+	case godata.ExpressionTokenGuid:
+		value = node.Token.Value
+	case godata.ExpressionTokenString:
+		// unquote
+		value = strings.Trim(node.Token.Value, "'")
+	default:
+		return "", unsupportedFilterError()
+	}
+	return value, nil
+}
+
+func (g Graph) isAppRoleAssignmentFilter(ctx context.Context, node *godata.ParseNode) (match bool, property string, filter string) {
+	if node.Token.Type != godata.ExpressionTokenLambdaNav {
+		return false, "", ""
+	}
+
+	if node.Children[0].Token.Type != godata.ExpressionTokenLiteral || node.Children[0].Token.Value != "appRoleAssignments" {
```

Comment:
It would imo be better to check for `len(node.Children)` in this function too. Otherwise the func will panic when it is misused.

Counts also for the `lambaParam` var introduced later.

Reply:
Yeah. Better safe than sorry. Though I think the godata Query Parser would already out on these kind of broken filters long before hitting any of the len() checks.